temporalio/activity.py

```diff
@@ -0,0 +1,287 @@
+
+    Retrieved inside an activity via :py:func:`info`.
+    """
+
+    activity_id: str
+    activity_type: str
+    attempt: int
+    heartbeat_details: Iterable[Any]
+    heartbeat_timeout: Optional[timedelta]
+    schedule_to_close_timeout: Optional[timedelta]
+    scheduled_time: datetime
+    start_to_close_timeout: Optional[timedelta]
+    started_time: datetime
+    task_queue: str
+    task_token: bytes
+    workflow_id: str
+    workflow_namespace: str
+    workflow_run_id: str
+    workflow_type: str
+    # TODO(cretz): Add headers, current_attempt_scheduled_time, retry_policy, and is_local
+    # TODO(cretz): Consider putting identity on here for "worker_id" for logger
+
+
+_current_context: contextvars.ContextVar[_Context] = contextvars.ContextVar("activity")
+
+
+@dataclass
+class _Context:
+    info: Callable[[], Info]
+    # This is optional because during interceptor init it is not present
+    heartbeat: Optional[Callable[..., None]]
+    cancelled_event: _CompositeEvent
+    worker_shutdown_event: _CompositeEvent
+    _logger_details: Optional[Mapping[str, Any]] = None
+
+    @staticmethod
+    def current() -> _Context:
+        context = _current_context.get(None)
+        if not context:
+            raise RuntimeError("Not in activity context")
+        return context
+
+    @staticmethod
+    def set(context: _Context) -> None:
+        _current_context.set(context)
+
+    @property
+    def logger_details(self) -> Mapping[str, Any]:
+        if self._logger_details is None:
+            info = self.info()
+            self._logger_details = {
+                "activity_id": info.activity_id,
+                "activity_type": info.activity_type,
+                "attempt": info.attempt,
+                "namespace": info.workflow_namespace,
+                "task_queue": info.task_queue,
+                "workflow_id": info.workflow_id,
+                "workflow_run_id": info.workflow_run_id,
+                "workflow_type": info.workflow_type,
+            }
+        return self._logger_details
+
+
+@dataclass
+class _CompositeEvent:
+    thread_event: threading.Event
+    # Async event only for async activities
+    async_event: Optional[asyncio.Event]
+
+    def set(self) -> None:
+        self.thread_event.set()
+        if self.async_event:
+            self.async_event.set()
+
+
+def in_activity() -> bool:
+    """Whether the current code is inside an activity.
+
+    Returns:
+        True if in an activity, False otherwise.
+    """
+    return not _current_context.get(None) is None
+
+
+def info() -> Info:
+    """Current activity's info.
+
+    Returns:
+        Info for the currently running activity.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().info()
+
+
+def heartbeat(*details: Any) -> None:
+    """Send a heartbeat for the current activity.
+
+    Raises:
+        RuntimeError: When not in an activity.
```

Comment:
OK, here's some notes on what I think I learned here:

* TypeScript
  * The heartbeat call does not throw and you will never know an activity cancel request happened unless you opt-in to knowing or are calling a cancellation aware call
  * A promise is resolved on the first cancellation request and is not reset if ignored/swallowed. So you can't, swallow the first cancel and wait for the next.
* Go
  * The heartbeat call does not throw, you leverage the Go context to know if cancelled
  * Like TypeScript, there is no way to swallow one cancellation and get notified about the next
  * Unlike TypeScript, there is nothing you can check to see if a cancellation was requested. Checking the context alone doesn't tell you it isn't, say, a worker stop.
* Java
  * The heartbeat call does throw, and properly resets itself for more cancellation requests if you swallow it (at least from what I am seeing)
  * Heartbeat is the only path for cancellation. There is nothing you can check or wait on to see if cancellation was requested

Some of those make sense in their async or sync environments. But Python is both, so what do we do in Python? Notes:

* Does `temporalio.activity.heartbeat` throw?
  * Thinking yes, but since it's asynchronously sent in the background, it can only throw on the _next_ call to heartbeat after the cancel is returned
  * This will have to throw a different error than `asyncio.CancelledError`, maybe called `ActivityCancelledError` or similar
    * I am thinking we want the cancellation reason on this too of which there are 4:
      * cancel requested
      * not found - gone from server, so the error is ignored anyways
      * timed out - being handled elsewhere, so the error will be ignored anyways
      * heartbeat failure - only really occurs if the details can't be converted on the previous heartbeat call
* Is `asyncio.Task.cancel` still used?
  * Thinking yes, but it's a different exception type and won't have the reason
    * Maybe make the activity cancelled error available for the reason?
* Should we allow resetting of the `is_cancelled` and `heartbeat` if the exception is swallowed
  * Thinking yes, but it's manual. So instead of like Java where the `heartbeat` call won't continue to throw on error, this will unless you manually reset the cancellation
  * The reason for doing this is to make the event simple
* This is all made more difficult by multiprocess support
  * It was simple unidirectional heartbeating, but now I've learned heartbeating has to be bidirectional to report errors

Still working on the best solution here...

Reply:
After discussion with @bergundy, we have resolved to...have further discussions heh. While heartbeat throwing may be ok, we have to discuss how that looks in local activities. Also need to discuss the confusion when your `await` call could throw one type of cancel _or_ your `heartbeat()` call could throw another type. I suspect this same concern will come up in .Net where you have async and non-async and both are fairly heavily used in the ecosystem.